[BE] Feat: 미완료 북마크 목록 조회 API 추가 및 사용하지 않는 v1 API 삭제#643

gabean13 opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->

### 1. api/v1/bookmarks/hearits API 삭제
관련 코드, 테스트 삭제했습니다.

### 2. query parameter filter 추가
queryParam로 filter=all 을 할 수 있도록 BookmarkFilter를 추가했습니다.
이때 url은 소문자여야한다고 생각, controller에서 받을때는 enum으로 받기 를 하고 싶어서
StringToBookmarkFilterConverter 를 ConverterWebConfig를 통해 등록했습니다.

그래서 filter=all 이면 BookmarkFilter=ALL로 매핑이되고,
만약 매핑에 실패하면 MethodArgumentTypeMismatchException 가 발생합니다.

## 📸 스크린샷 (선택)
> UI 변경이 있을 경우 첨부
<img width="1468" height="709" alt="image" src="https://github.com/user-attachments/assets/4b36e46c-c636-457f-8ec3-61680a949ed7" />


## 🧪 테스트 방법 (선택)
<!-- 코드 리뷰하는 팀원이 돌려볼 수 있도록 작성해주기 
ex. ReservationTest의 XXXX 테스트메서드로 확인 
ex. 어떤화면의 어떤기능에서 확인 -->

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->

### BookmarkFilter, StringToBookmarkFilterConverter, ConverterWebConfig 의 패키지 위치를 어디로 해야할지?
현재 BookmarkFilter가 controller, service에서 사용이 되는데, dto로 보기에는 어느정도 책임이 있는 것 같고 도메인으로 보기에는 무리이고..
해서 애매합니다...
converter랑 config도 어떻게 해야할지 다같이 협의하면 좋을 것 같습니다.
제 생각에는 converter는 bookmarkfilter랑 같은 패키지에 넣고, 
webconfig는 앞으로 나오는 query parmeter들 변환하는 클래스들 다 같이 넣게 config패키지를 도메인이랑 동일한 위치에 만들어서 관리하면 좋을 것 같다고 생각을 했습니다.

### nested 적극 활용
저희 테스트가 점점 뚱뚱해지는데 api 별로 묶는거 어떻게 생각하시나욧?
한꺼번에 적용하기보다 짤때마다 하나하나씩 점차 적용시켜나가면 좋을 것 같습니다.

서비스까지는 안해도 될 것 같고 controller test랑 integration test는 하면 버저닝할때랑 응답별로 작성할때 유용할 것 같습니다.
제 BookmarkController test보고 괜찮다고 생각하신다면 Integeration test에도 적용하겠ㅅ습니다.

좋은 의견 부탁합니다.

### isMemberOnly 이거 언제 적용할건지
응답이 추가되는거라 깨지진 않을 것 같은데 지금 적용할건지 얘기하고 싶습니다.
변수명도!

## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->

#643 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- 신기능
  - 북마크 목록에 필터/정렬 파라미터 지원: filter=all|unfinished, sort=createdAt,asc|desc
  - /api/v1/bookmarks 및 /api/v2/bookmarks/hearits에서 동일한 V2 응답 포맷 제공
  - 쿼리파라미터 변환기 추가로 filter/sort 문자열을 직접 전달 가능

- 버그 수정
  - 결과가 없을 때 일관된 빈 페이징 응답 반환 (빈 리스트, 페이지 정보 0)

- 리팩터
  - V1 목록 API 통합·파라미터화 및 미완료 전용 엔드포인트 제거
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gabean13 committed:
[BE] feat: BookmarkFilter enum 및 테스트 추가 #643

gabean13 committed:
[BE] feat: StringToBookmarkFilterConverter 추가 및 WebConfig에 등록 #643

- 문자열을 BookmarkFilter로 변환하는 StringToBookmarkFilterConverter 클래스 추가
- WebConfig에서 StringToBookmarkFilterConverter를 FormatterRegistry에 등록

gabean13 committed:
[BE] feat: 멤버의 미완료 북마크 조회 기능 추가 #643

- 북마크 중 청취 미완료 히어릿 조회를 위한 findUnfinishedByMemberOrderByRecent 쿼리 추가
- findUnfinishedByMemberOrderByRecentTest 테스트 케이스 작성

gabean13 committed:
[BE] feat: 북마크 조회 시 필터링 기능 추가 #643

- 북마크 필터링 로직 추가로 모든 북마크와 미완료 북마크 조회 가능
- 비회원 조회 시 빈 값 반환

gabean13 committed:
[BE] feat: 비어있는 PagedResponse를 반환하는 로직 추가 #643

- page가 비어있을 경우 empty PagedResponse 반환하도록 로직 추가

gabean13 committed:
[BE] feat: 북마크 필터링 조회 API 추가 #643

- 북마크 조회 API에 필터링 기능 추가 (all, unfinished 등)
- 북마크 목록 조회 API를 V1, V2로 분리하여 구현

gabean13 committed:
[BE] feat: /api/*/bookmarks` Endpoint에 대해 CORS 허용 #643

gabean13 committed:
[BE] chore: 테스트 데이터 초기화를 위한 AUTO_INCREMENT 리셋 추가 #643

- `dbclean.sql` 파일에 AUTO_INCREMENT 초기화를 위한 SQL 구문 추가

gabean13 committed:
[BE] refactor: V1 북마크 조회 API 및 관련 테스트 제거 #643

- `/api/v1/bookmarks/hearits` 엔드포인트 제거
- 관련 테스트 코드 삭제

gabean13 committed:
[BE] refactor: WebConfig 클래스 이름을 ConverterWebConfig으로 변경 #643

gabean13 committed:
[BE] fix: 기존 API 삭제로 유효한 토큰 테스트 로직 변경 #643

Byesol commented on backend/app/src/main/java/com/onair/hearit/app/bookmark/presentation/BookmarkController.java:
```diff
@@ -27,38 +25,34 @@ public class BookmarkController {
+            @RequestParam(name = "filter", defaultValue = "all") BookmarkFilter filter,
             @RequestParam(name = "page", defaultValue = "0") int page,
             @RequestParam(name = "size", defaultValue = "20") int size,
             @AuthenticationPrincipal RequestUser requestUser) {
```
현재 정렬순서는 createdAt DESC 고정인데 정렬 기준도 parameter로 빼서 확장성을 늘리는 것도 좋아보여요

gabean13 replied:
까악 햇씁니다

Byesol commented on backend/app/src/main/java/com/onair/hearit/app/bookmark/dto/param/BookmarkFilter.java:
```diff
@@ -0,0 +1,27 @@
+        return Arrays.stream(values())
+                .filter(value -> value.name.equalsIgnoreCase(queryParam))
+                .findFirst()
+                .orElseThrow(IllegalArgumentException::new);
```
여기서 ileegalargument를 던지면 MethodArgumentTypeMismatchException로 잡히긴 하겠지만 

저희가 관리하는 invalidiinputexcepction 와 구체적인 메시지를 던지는 건 어떨까요?

gabean13 replied:
Converter내부에서 에러 발생 시 spring이 methodargumentexception으로 변환해서 던지기 때문에 이 방식은 어려울 것 같습니다ㅜ

JO-eusan replied:
!! 새롭게 알아갑니다.^_^

gabean13 committed:
[BE] fix: 북마크 필터 조건에 optional 속성 추가 #643

Byesol reviewed (commented):
1. BookmarkFilter는 비즈니스 로직과 관련된 조건이므로 application 계층,(패키지지)
Converter는 외부의 String 데이터를  비즈니스 로직 관련된 BookmarkFilter 로 변환시키고 있으므로 presentation 패키지에 들어가야된다고 생각해용. 
 
 그리고 config를 도메인과 같은 위치에 넣자는 게 core 모듈에 넣자는 말인가요?
 저는 app / config 에 넣는 게 낫다고 생각해요.
 
  앞으로 나오는 query parmeter들 변환하는 클래스들 다 같이 넣앞으로 나오는 query parmeter들 변환하는 클래스들 다 같이 넣는 용도면 app/config 면 충분하다고 생각해요.
  
  core 모듈의 순수성을 유지하기 위해 app 모듈의 app/config 로 관리하는 게 낫다고 생각했는데
  이미 core/ config 에 SecurityBeanConfig가 있긴 하네용.
  WebMvcConfigurer 를 이용하는 건 결국 spring mvc 기술에 강하게 의존하는 거라, core에 위치하면 안 된다는 생각이 들긴 하는데 이에 대한 의견들이 궁금해요.


2. 가콩의 테스트들을 보니 nested 는 쓰는 게 훨씬  깔끔해보이고 좋은 것 같아요! 

그런데 기존 테스트를 전부 해당 방식으로 교체한는 건 또 다른 task 가 될 수 있으니 해당 부분 수정할 일 생길때마다 점진적으로 nested로 교체하는 것도 괜찮아보여요!

3. isMemberOnly로 인해 어차피 응답이 깨지는 게 아니니 , 미리 api 응답값으로 줘버리는 게 괜찮을 것 같습니다!

JO-eusan commented on backend/app/src/main/java/com/onair/hearit/app/bookmark/application/BookmarkService.java:
```diff
@@ -33,17 +31,36 @@ public class BookmarkService {
+    public Page<BookmarkHearitResponseV2> getBookmarkHearitsV2(UserInfo userInfo,
+                                                               PagingRequest pagingRequest,
+                                                               BookmarkFilter filter) {
+        Member member = getMemberByUserInfo(userInfo);
```
이 부분도 사용하지 않는 api지만, 
지금 ApiSecurityConfig에서 화이트 리스트로 `/bookmarks`를 넣고 있어서 예외 메시지가 달라질 수도 있을 것 같아요. (just 추측)
위에처럼 `isGuest()`일때 빈 리스트를 반환하는 처리를 해줘야하지 않을까요?

gabean13 replied:
화이트리스트로 넣어주고 있는 url은 "/api/*/bookmarks" 이고,  이 메서드에서 사용하는 url은 @GetMapping("/api/v2/bookmarks/hearits")이라서 현재 guest인 경우 아예 이 메서드까지 도달하지 않습니닷

JO-eusan commented on backend/app/src/main/java/com/onair/hearit/app/bookmark/application/BookmarkService.java:
```diff
@@ -33,17 +31,36 @@ public class BookmarkService {
+        return toBookmarkHearitResponse(projections);
+    }
+
+    private Page<BookmarkWithPlayingHistoryProjection> getFilteredBookmarkHearits(Member member,
```
저는 필터링 조건을 바로 쿼리에 사용할 수 있도록 처리를 해줬는데요,
가콩도 하나의 쿼리로 Controller에서 `required=false`를 통해 null이 들어오게 한 후, isFinished라는 북마크의 필드만 enum으로 추가하는건 어떨까요? 
이렇게 switch 문으로 하게되면 다른 필드로 필터링을 하고 싶을때 리팩터링하기 쉽지 않을 것 같아요!

예상 쿼리는 다음과 같습니다.
```sql
SELECT b AS bookmark, ph AS playingHistory
FROM Bookmark b
JOIN FETCH b.hearit h
JOIN FETCH h.category c
LEFT JOIN PlayingHistory ph 
  ON ph.hearitId = h.id AND ph.memberId = :memberId
WHERE b.member.id = :memberId
AND (:isFinished IS NULL OR ph.isFinished = :isFinished)
ORDER BY b.createdAt DESC

```

gabean13 replied:
반영 했습니다

JO-eusan commented on backend/app/src/main/java/com/onair/hearit/app/bookmark/dto/param/BookmarkFilter.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.Arrays;
+
+public enum BookmarkFilter {
```
이 Enum 클래스 자체는 DTO의 성격이 있다고 생각합니다.
dto 내에 파라미터 조건들을 넣어두는 패키지를 넣어서 관리하면 어떨까요?

ex) bookmark\dto\param\BookmarkFilter.class

JO-eusan commented on backend/app/src/main/java/com/onair/hearit/app/bookmark/StringToBookmarkFilterConverter.java:
```diff
@@ -0,0 +1,11 @@
+
+import org.springframework.core.convert.converter.Converter;
+
+public class StringToBookmarkFilterConverter implements Converter<String, BookmarkFilter> {
```
BookmarkFilterConverter라고 표현해도 될 것 같은데 가콩은 어떻게 생각하시나요?
converter.convert(String source)에서 파라미터의 타입을 보고 추론이 가능하다고 생각합니다.

gabean13 replied:
좋습니두

JO-eusan commented on backend/app/src/main/java/com/onair/hearit/app/bookmark/presentation/BookmarkController.java:
```diff
@@ -27,38 +25,34 @@ public class BookmarkController {
+                requestUser.getUserInfo(),
+                pagingRequest,
+                BookmarkFilter.ALL);
+        return ResponseEntity.ok(PagedResponse.from(bookmarkHearits));
```
저희 컨벤션으로 DTO 변환은 Service에서 하기로 해서 기존 `from` 호출부를 보면 모두 Service에요! Service 계층에서 변환해서 받아오는건 어떨까요?

gabean13 replied:
이거 v1, v2 때문에 일케 했었는데 v1 삭제했으니 사나 말대로 바꿀게용

JO-eusan commented on backend/app/src/test/java/com/onair/hearit/app/bookmark/presentation/BookmarkControllerTest.java:
```diff
@@ -58,290 +56,383 @@ void readBookmarkHearitsV1_OK() throws Exception {
+                            resource(ResourceSnippetParameters.builder()
+                                    .tag("Bookmark API")
+                                    .summary("북마크 목록 조회 V1")
+                                    .description("로그인한 사용자가 북마크한 히어릿 목록을 `filter(" + String.join(", ",
```
이거 filter 조건을 description보다 queryParameters 부분에 자세하게 적는건 어떻게 생각하시나요?

gabean13 replied:
좋아욤 햇슴니다

JO-eusan reviewed (changes requested):
가콩 연휴에도 고생하셨어요ㅠㅠ
몇개 리뷰 남겨두었습니다!

gabean13 committed:
[BE] feat: 북마크 정렬 파라미터 처리 기능 추가 #643

- BookmarkSort 및 BookmarkSortConverter 클래스 추가로 정렬 파라미터 처리 구현
- 기본 정렬 방향을 DESC로 설정

gabean13 committed:
[BE] refactor: 북마크 필터 패키지 이동 #643

- BookmarkFilter 관련 클래스들을 적합한 dto 하위 패키지로 이동
- StringToBookmarkFilterConverter 클래스 이름을 BookmarkFilterConverter로 변경

yuyoungrhee commented on backend/core/src/testFixtures/resources/dbclean.sql:
```diff
@@ -12,3 +12,16 @@ TRUNCATE TABLE hearit_source;
 TRUNCATE TABLE explore_score;
 TRUNCATE TABLE playing_history;
 SET FOREIGN_KEY_CHECKS = 1;
+
+-- AUTO_INCREMENT 리셋 추가
+ALTER TABLE member AUTO_INCREMENT = 1;
+ALTER TABLE hearit AUTO_INCREMENT = 1;
+ALTER TABLE category AUTO_INCREMENT = 1;
+ALTER TABLE playing_history AUTO_INCREMENT = 1;
+ALTER TABLE bookmark AUTO_INCREMENT = 1;
+ALTER TABLE keyword AUTO_INCREMENT = 1;
+ALTER TABLE admin AUTO_INCREMENT = 1;
+ALTER TABLE refresh_token AUTO_INCREMENT = 1;
+ALTER TABLE recommend_hearit AUTO_INCREMENT = 1;
+ALTER TABLE hearit_source AUTO_INCREMENT = 1;
+ALTER TABLE explore_score AUTO_INCREMENT = 1;
```
id에 대해 의존하는 테스트가 좋다고 생각하진 않는데, auto increment라는 id 생성방식 자체의 구현에 대해서 의존하게 돼서 로직 + db의 id 생성 규칙 이렇게 검증하게 되기 때문입니다. 혹시 정확히 어떤 테스트가 필요했던 건가요?

yuyoungrhee reviewed (commented):
깔끔한데엽 수고하셨습니다
궁금한 부분 코멘트만 하나 달게요

- nested 적극활용에 대해서 적극 찬성입니다. 마침 사나 리뷰에도 달고왔더염 ㅋㅋ
- isMemberOnly는 벡터가 말한 것처럼 그냥 할일다끝낸사람이 혹시 시간남으면 빨리 이슈파서 추가해버려도 괜찮을 듯요 일단 더준다고 깨지진않아서

gabean13 committed:
[BE] refactor: 북마크 조회 API 통합 및 필터링 기능 개선 #643

- `findAllByMemberOrderByRecent`와 `findUnfinishedByMemberOrderByRecent` 메서드를 `findFilteredByMember`로 통합
- 필터링 조건 `isFinished`를 추가하여 완료/미완료 상태 조회 지원

gabean13 committed:
[BE] feat: 북마크 API에 정렬 기능 추가 #643

- 북마크 조회 서비스 및 컨트롤러에 정렬 로직 적용

gabean13 committed:
[BE] feat: 북마크 목록 조회 시 정렬 파라미터 추가 #643

- V1, V2 북마크 조회 API에 정렬(`sort`) 파라미터 추가
- 정렬 조건을 위한 `BookmarkSort` 추가 및 관련 테스트 수정
- API 문서에 정렬 조건 설명 필드 추가

gabean13 committed:
[BE] Merge branch 'develop-be' into feat/#643-unfinished-bookmark #643

# Conflicts:
#	backend/app/src/test/java/com/onair/hearit/app/auth/config/ApiSecurityConfigTest.java

gabean13 committed:
[BE] refactor: dbclean.sql에서 AUTO_INCREMENT 리셋 구문 제거 #643

gabean13 committed:
[BE] refactor: 테스트 데이터 생성 로직을 TestFixture로 정리 및 클래스 생성자 호출 순서 변경 #643

- 테스트 코드에서 중복된 `createHearitWith` 메서드 제거 및 TestFixture로 대체
- `GuestExploreScoreProcessor`, `MemberExploreScoreProcessor` 생성자 내부 호출 순서 변경

gabean13 committed:
[BE] refactor: BookmarkSortConverter에서 정렬 처리 로직 개선 #643

- 입력 값에서 빈 문자열 필터링 로직 추가
- single-value, two-value 정렬 조건 처리 방식 개선
- 유효하지 않은 입력 값 예외 처리 로직 추가

Byesol reviewed (approved):
수고했어용 가콩. 어푸할께영

JO-eusan reviewed (approved):
리뷰 반영 내용 확인했습니다! Approve할게요~
